### PR TITLE
Dev container - dev tools enhancement

### PR DESCRIPTION
**Pull request by KYLChiu**
Aims to improve developer experience with vscode devcontainer:

* Use ruff as a drop in replacement for isort/black. Now devcontainer auto formats your code and also auto sorts your import on save.
* Adds pre-commit and pre-push. When running git commit, it auto runs our checks on basic stuff, like no case sensitive filename crashes. When running git push, it auto runs ruff formatting/linting. Ruff automatically formats your files to ensure its happy.  pre-push also checks other things (dockerfile linting). 
* Fix some minor complaints on import from previous PR.

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -22,13 +22,24 @@
...
+            ],
+            "settings": {
+                "[python]": {
+                    "editor.formatOnSave": true,
+                    "editor.defaultFormatter": "charliermarsh.ruff",
+                    "editor.codeActionsOnSave": {
+                        "source.organizeImports": "explicit"
+                    }
+                }
+            }
```
Comment (richardhunghhw) [inline .devcontainer/devcontainer.json:42]: You should add the settings files in `.vscode/settings.json` instead

File: sporkfish/evaluator.py [hunk 2]
```diff
@@ -234,9 +232,8 @@ def evaluate(self, board: Board) -> float:
         # Assumes:
         # - Chess board implements A1 as first element, H8 as last
         # - Piece square table implements A8 as first element, H1 as last element
-        flip: Callable[[int, int, chess.Color]] = (
-            lambda square, flipped_square, color: square if not color else flipped_square
-        )
+        def flip(square: int, flipped_square: int, color: chess.Color) -> int:
+            return square if not color else flipped_square
```
Comment (ccjeremylo) [inline sporkfish/evaluator.py:236]: Curious: why decide to use a normal function in the end? (done by ruff automatically?)

File: tests/perf_helper.py [hunk 3]
```diff
@@ -1,3 +1,7 @@
+import cProfile
+import os
+import pstats
+import sys
```
Comment (ccjeremylo) [inline tests/perf_helper.py:4]: nicee - tyty, was this done manually or via ruff?

File: .github/workflows/dev.yml [hunk 4]
```diff
@@ -32,10 +32,7 @@ jobs:
...
-      - name: Check code format (black)
-        run: docker exec sporkfish-container black --check .
-      - name: Check imports format (isort)
-        run: docker exec sporkfish-container isort --check --profile black .
-        
-     
-      
\ No newline at end of file
+      - name: Check code and imports format (ruff)
+        run: docker exec sporkfish-container ruff check .
```
Comment (ccjeremylo) [inline .github/workflows/dev.yml:36]: Sanity checking: Is my understanding correct that the reasons for replacing black/isort with ruff are (based on [this](https://www.packetcoders.io/whats-the-difference-black-vs-ruff/#:~:text=The%20choice%20between%20Black%20and,Ruff%20is%20a%20great%20choice.)):

- ruff is faster (not like black was slow tbf?)
- ruff does what black and isort would do (formatting), as well as [linting](https://code.visualstudio.com/docs/python/linting#:~:text=Linting%20is%20distinct%20from%20Formatting,only%20restructures%20how%20code%20appears.)

File: .gitignore [hunk 5]
```diff
@@ -158,6 +158,8 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+
+# API token
```
Comment (ccjeremylo) [inline .gitignore:162]: Tyty

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
+repos:
```
Comment (ccjeremylo) [inline .pre-commit-config.yaml:1]: Again sanity checking: this file basically does the [pre-commit](https://pre-commit.com/) checks for us, a.k.a automating the trivial checks we run before submitting?

On a high level, this is added to improve development experience?

File: scripts/container-postCreate.sh [hunk 7]
```diff
@@ -3,4 +3,6 @@
 # Run pip install
 pip install -r requirements.txt
 
+pre-commit install
```
Comment (ccjeremylo) [inline scripts/container-postCreate.sh:6]: Sorry not quite clear to me when will this script be executed? what does post create mean?

File: sporkfish/evaluator.py [hunk 8]
```diff
@@ -1,10 +1,8 @@
 from enum import Enum
-from typing import Callable
 
 import chess
 
 from .board.board import Board
-from .configurable import Configurable
```
Comment (ccjeremylo) [inline sporkfish/evaluator.py:7]: Again, want to check - did you do these manually or was it done automatically by ruff/pre-commit magic?

**Review by ccjeremylo — COMMENTED** [review]
Asking some very basic questions, want to make sure I understand how this works - plz bare with me

File: .github/workflows/dev.yml (hunk 4 above)
Comment (KYLChiu) [reply .github/workflows/dev.yml:36]: The latter, plus I couldn't get auto isort to work

File: .pre-commit-config.yaml (hunk 6 above)
Comment (KYLChiu) [reply .pre-commit-config.yaml:1]: Yes exactly.

Discussed offline, pre-commit currently does mypy checks and blocks commits - which can be fairly painful. Will try pre-push instead.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    hooks:
+      # Run the linter.
+      - id: ruff
+        args: [ --fix ]
+      # Run the formatter.
+      - id: ruff-format
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: "v1.8.0"
+    hooks:
+      - id: mypy
```
Comment (ccjeremylo) [inline .pre-commit-config.yaml:33]: Discussed offline - I am not in favour of having `mypy` in pre-commit or pre-post. Sometimes I would like to pass on some unpolished work to others by pushing, or I simply want to push some unfinished code to remote branch so it is safe.

Not sure I would want to have `mypy` blocking me to push, especially if the code is not even at a state for type hints. 

Thoughts?

File: sporkfish/evaluator.py (hunk 2 above)
Comment (KYLChiu) [reply sporkfish/evaluator.py:236]: Ruff agrees with you - doesn't like lambdas. though type hints manually added

File: tests/perf_helper.py (hunk 3 above)
Comment (KYLChiu) [reply tests/perf_helper.py:4]: Manually for this one

File: .pre-commit-config.yaml (hunk 9 above)
Comment (KYLChiu) [reply .pre-commit-config.yaml:33]: Will try move to pre-push.

File: sporkfish/evaluator.py (hunk 8 above)
Comment (KYLChiu) [reply sporkfish/evaluator.py:7]: magic

File: scripts/container-postCreate.sh (hunk 7 above)
Comment (KYLChiu) [reply scripts/container-postCreate.sh:6]: After the container is built, this is run - please see the comment in `devcontainer.json`

File: .pre-commit-config.yaml (hunk 9 above)
Comment (KYLChiu) [reply .pre-commit-config.yaml:46]: Ok, fair enough. I'll remove this. I would argue that people can use git push --no-verify to pass the checks but I'm aware not everyone is using CLI.

**Review by richardhunghhw — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 6 above)
Comment (ccjeremylo) [reply .pre-commit-config.yaml:1]: Ty!!

File: sporkfish/searcher/minimax.py [hunk 10]
```diff
@@ -228,9 +228,6 @@ def _iterative_deepening(
         for depth in range(1, self._searcher_config.max_depth + 1):
             new_board = copy.deepcopy(board)
 
-            alpha = -float("inf")
```
Comment (ccjeremylo) [inline sporkfish/searcher/minimax.py:231]: Not too sure why we don't need these two lines now, but given consistency tests still pass, seems good to me.

**Review by ccjeremylo — APPROVED** [review]
LGTM, excited to try this 😃

**Pull request merged**